src/perception/tag_detector/tag_detector_thresh.cpp

```diff
@@ -0,0 +1,89 @@
+#include "tag_detector.hpp"
+
+#include <cmath>
+#include <vector>
+#include <execution>
+
+#include <sensor_msgs/image_encodings.h>
+#include <opencv2/opencv.hpp>
+
+namespace mrover {
+    /**
+     * @brief Threshold input image using adaptive thresholding
+     */
+    static void threshold(cv::InputArray _in, cv::OutputArray _out, int winSize, double constant) {
+
+        CV_Assert(winSize >= 3);
+        if(winSize % 2 == 0) winSize++; // win size must be odd
+        cv::adaptiveThreshold(_in, _out, 255, cv::ADAPTIVE_THRESH_MEAN_C, cv::THRESH_BINARY_INV, winSize, constant);
+    }
+
+    /**
+     * @brief Convert input image to gray if it is a 3-channels image
+     */
+    void convertToGrey(cv::InputArray _in, cv::OutputArray _out) {
+
+        CV_Assert(_in.type() == CV_8UC1 || _in.type() == CV_8UC3);
+
+        if(_in.type() == CV_8UC3)
+            cvtColor(_in, _out, cv::COLOR_BGR2GRAY);
+        else
+            _in.copyTo(_out);
+    }
+
+    /**
+     * @brief Initial steps on finding square candidates
+     */
+    void runAdaptiveThresh(cv::Mat &grey, const cv::Ptr<cv::aruco::DetectorParameters> &params) {
+
+        CV_Assert(params->adaptiveThreshWinSizeMin >= 3 && params->adaptiveThreshWinSizeMax >= 3);
+        CV_Assert(params->adaptiveThreshWinSizeMax >= params->adaptiveThreshWinSizeMin);
+        CV_Assert(params->adaptiveThreshWinSizeStep > 0);
+
+        // number of window sizes (scales) to apply adaptive thresholding
+        int nScales =  (params->adaptiveThreshWinSizeMax - params->adaptiveThreshWinSizeMin) /
+                        params->adaptiveThreshWinSizeStep + 1;
+
+        ////for each value in the interval of thresholding window sizes
+        parallel_for_(cv::Range(0, nScales), [&](const cv::Range& range) {
```

Comment:
I will have to think about this even works... it seems like if ran in parallel they would scribble all over eachother...

@niwhsa9 any clues? This is pulled from opencv source code for aruco detection